File: client-scalaz/src/main/scala/org/openapitools/client/api/GameLevel.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime

import java.time.OffsetDateTime

import GameLevel._

case class GameLevel (
  id: Option[Long],
active: Option[Boolean],
valid: Option[Boolean],
newOwnerId: Option[Long],
gameDifficulty: Option[GameDifficulty],
name: Option[String],
image: Option[Asset],
icon: Option[Asset],
description: Option[String],
startDate: Option[OffsetDateTime],
endDate: Option[OffsetDateTime],
randomizeGameObjects: Option[Boolean],
owner: Option[Account],
version: Option[AppVersion],
notes: Option[List[Note]],
noteCount: Option[Long],
likes: Option[List[YayOrNay]],
likeCount: Option[Long],
dislikeCount: Option[Long],
playCount: Option[Long],
downloadCount: Option[Long],
gameObjectCount: Option[Integer],
publicPermissions: Option[Permissions],
visibility: Option[Visibility],
userPermissions: Option[List[UserPermissions]],
flags: Option[List[Flag]],
flagCount: Option[Long],
flagThreshold: Option[FlagThreshold],
application: Option[Application],
assignMission: Option[Boolean],
authorOverride: Option[String],
splashTitle: Option[String],
splashMessage: Option[String],
winnerMessage: Option[String],
tutorial: Option[Tutorial],
approvalStatus: Option[ApprovalStatus],
nextLevel: Option[GameLevel],
offer: Option[Offer],
balance: Option[Double],
levelNumber: Option[Integer],
points: Option[Long],
ticketType: Option[String],
ticketCount: Option[Long],
allocateTickets: Option[Boolean],
applicationTitle: Option[String],
ownerUsername: Option[String],
likableNotificationModels: Option[List[Map[String, Any]]],
notableNotificationModels: Option[List[Map[String, Any]]],
published: Option[Boolean],
contentName: Option[String],
defaultThreshold: Option[Long],
contentAsset: Option[Asset],
`contentType`: Option[String])

object GameLevel {
  import DateTimeCodecs._
  sealed trait GameDifficulty
  case object VERYEASY extends GameDifficulty
  case object EASY extends GameDifficulty
  case object MEDIUM extends GameDifficulty
  case object HARD extends GameDifficulty
  case object VERYHARD extends GameDifficulty

  object GameDifficulty {
    def toGameDifficulty(s: String): Option[GameDifficulty] = s match {
      case "VERYEASY" => Some(VERYEASY)
      case "EASY" => Some(EASY)
      case "MEDIUM" => Some(MEDIUM)
      case "HARD" => Some(HARD)
      case "VERYHARD" => Some(VERYHARD)
      case _ => None
    }

    def fromGameDifficulty(x: GameDifficulty): String = x match {
      case VERYEASY => "VERYEASY"
      case EASY => "EASY"
      case MEDIUM => "MEDIUM"
      case HARD => "HARD"
      case VERYHARD => "VERYHARD"
    }
  }

  implicit val GameDifficultyEnumEncoder: EncodeJson[GameDifficulty] =
    EncodeJson[GameDifficulty](is => StringEncodeJson(GameDifficulty.fromGameDifficulty(is)))

  implicit val GameDifficultyEnumDecoder: DecodeJson[GameDifficulty] =
    DecodeJson.optionDecoder[GameDifficulty](n => n.string.flatMap(jStr => GameDifficulty.toGameDifficulty(jStr)), "GameDifficulty failed to de-serialize")
  sealed trait Visibility
  case object PUBLIC extends Visibility
  case object `PRIVATE` extends Visibility
  case object FRIENDS extends Visibility

  object Visibility {
    def toVisibility(s: String): Option[Visibility] = s match {
      case "PUBLIC" => Some(PUBLIC)
      case "`PRIVATE`" => Some(`PRIVATE`)
      case "FRIENDS" => Some(FRIENDS)
      case _ => None
    }

    def fromVisibility(x: Visibility): String = x match {
      case PUBLIC => "PUBLIC"
      case `PRIVATE` => "`PRIVATE`"
      case FRIENDS => "FRIENDS"
    }
  }

  implicit val VisibilityEnumEncoder: EncodeJson[Visibility] =
    EncodeJson[Visibility](is => StringEncodeJson(Visibility.fromVisibility(is)))

  implicit val VisibilityEnumDecoder: DecodeJson[Visibility] =
    DecodeJson.optionDecoder[Visibility](n => n.string.flatMap(jStr => Visibility.toVisibility(jStr)), "Visibility failed to de-serialize")
  sealed trait ApprovalStatus
  case object PENDING extends ApprovalStatus
  case object REJECTED extends ApprovalStatus
  case object APPROVED extends ApprovalStatus
  case object FEATURED extends ApprovalStatus

  object ApprovalStatus {
    def toApprovalStatus(s: String): Option[ApprovalStatus] = s match {
      case "PENDING" => Some(PENDING)
      case "REJECTED" => Some(REJECTED)
      case "APPROVED" => Some(APPROVED)
      case "FEATURED" => Some(FEATURED)
      case _ => None
    }

    def fromApprovalStatus(x: ApprovalStatus): String = x match {
      case PENDING => "PENDING"
      case REJECTED => "REJECTED"
      case APPROVED => "APPROVED"
      case FEATURED => "FEATURED"
    }
  }

  implicit val ApprovalStatusEnumEncoder: EncodeJson[ApprovalStatus] =
    EncodeJson[ApprovalStatus](is => StringEncodeJson(ApprovalStatus.fromApprovalStatus(is)))

  implicit val ApprovalStatusEnumDecoder: DecodeJson[ApprovalStatus] =
    DecodeJson.optionDecoder[ApprovalStatus](n => n.string.flatMap(jStr => ApprovalStatus.toApprovalStatus(jStr)), "ApprovalStatus failed to de-serialize")

  implicit val GameLevelCodecJson: CodecJson[GameLevel] = CodecJson.derive[GameLevel]
  implicit val GameLevelDecoder: EntityDecoder[GameLevel] = jsonOf[GameLevel]
  implicit val GameLevelEncoder: EntityEncoder[GameLevel] = jsonEncoderOf[GameLevel]
}

File: client-clojure/src/sirqul_io_t_platform/specs/game_level.clj
(ns sirqul-io-t-platform.specs.game-level
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.asset :refer :all]
            [sirqul-io-t-platform.specs.asset :refer :all]
            [sirqul-io-t-platform.specs.account :refer :all]
            [sirqul-io-t-platform.specs.app-version :refer :all]
            [sirqul-io-t-platform.specs.note :refer :all]
            [sirqul-io-t-platform.specs.yay-or-nay :refer :all]
            [sirqul-io-t-platform.specs.permissions :refer :all]
            [sirqul-io-t-platform.specs.user-permissions :refer :all]
            [sirqul-io-t-platform.specs.flag :refer :all]
            [sirqul-io-t-platform.specs.flag-threshold :refer :all]
            [sirqul-io-t-platform.specs.application :refer :all]
            [sirqul-io-t-platform.specs.tutorial :refer :all]
            [sirqul-io-t-platform.specs.game-level :refer :all]
            [sirqul-io-t-platform.specs.offer :refer :all]
            [sirqul-io-t-platform.specs.asset :refer :all]
            )
  (:import (java.io File)))


(def game-level-data
  {
   (ds/opt :id) int?
   (ds/opt :active) boolean?
   (ds/opt :valid) boolean?
   (ds/opt :newOwnerId) int?
   (ds/opt :gameDifficulty) string?
   (ds/opt :name) string?
   (ds/opt :image) asset-spec
   (ds/opt :icon) asset-spec
   (ds/opt :description) string?
   (ds/opt :startDate) inst?
   (ds/opt :endDate) inst?
   (ds/opt :randomizeGameObjects) boolean?
   (ds/opt :owner) account-spec
   (ds/opt :version) app-version-spec
   (ds/opt :notes) (s/coll-of note-spec)
   (ds/opt :noteCount) int?
   (ds/opt :likes) (s/coll-of yay-or-nay-spec)
   (ds/opt :likeCount) int?
   (ds/opt :dislikeCount) int?
   (ds/opt :playCount) int?
   (ds/opt :downloadCount) int?
   (ds/opt :gameObjectCount) int?
   (ds/opt :publicPermissions) permissions-spec
   (ds/opt :visibility) string?
   (ds/opt :userPermissions) (s/coll-of user-permissions-spec)
   (ds/opt :flags) (s/coll-of flag-spec)
   (ds/opt :flagCount) int?
   (ds/opt :flagThreshold) flag-threshold-spec
   (ds/opt :application) application-spec
   (ds/opt :assignMission) boolean?
   (ds/opt :authorOverride) string?
   (ds/opt :splashTitle) string?
   (ds/opt :splashMessage) string?
   (ds/opt :winnerMessage) string?
   (ds/opt :tutorial) tutorial-spec
   (ds/opt :approvalStatus) string?
   (ds/opt :nextLevel) game-level-spec
   (ds/opt :offer) offer-spec
   (ds/opt :balance) float?
   (ds/opt :levelNumber) int?
   (ds/opt :points) int?
   (ds/opt :ticketType) string?
   (ds/opt :ticketCount) int?
   (ds/opt :allocateTickets) boolean?
   (ds/opt :applicationTitle) string?
   (ds/opt :ownerUsername) string?
   (ds/opt :likableNotificationModels) (s/coll-of (s/map-of string? any?))
   (ds/opt :notableNotificationModels) (s/coll-of (s/map-of string? any?))
   (ds/opt :published) boolean?
   (ds/opt :contentName) string?
   (ds/opt :defaultThreshold) int?
   (ds/opt :contentAsset) asset-spec
   (ds/opt :contentType) string?
   })

(def game-level-spec
  (ds/spec
    {:name ::game-level
     :spec game-level-data}))
